include/bitcoin/bitcoin/math/ring_signature.hpp

```diff
@@ -0,0 +1,84 @@
+/**
+ * Copyright (c) 2011-2018 libbitcoin developers (see AUTHORS)
+ *
+ * This file is part of libbitcoin.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#ifndef LIBBITCOIN_RING_SIGNATURE_HPP
+#define LIBBITCOIN_RING_SIGNATURE_HPP
+
+#include <bitcoin/bitcoin/math/elliptic_curve.hpp>
+
+namespace libbitcoin {
+
+/**
+ * key_rings is a collection of public key rings.
+ * Each ring contains several public keys.
+ * To create a valid signature, you must use at least one valid private
+ * key from each ring.
+ * That is given three rings [{A, B, C}, {D, E, F}, {X, Y}], then to
+ * create a valid signature, we must use a private key from each of those
+ * sets.
+ * For example A and E and X. We can summarize this operation as:
+ * (A or B or C) and (D or E or F) and (X or Y)
+ */
+typedef std::vector<point_list> key_rings;
+
+/**
+ * A borromean ring signature.
+ *
+ * theta = {e_0, s_i_j : 0 <= i <= n, 0 <= j <= m_i}
+ */
+struct ring_signature
+{
+    typedef std::vector<secret_list> s_values_type;
+
+    ec_secret e;
+    s_values_type s;
+};
```

Comment:
Use existing typedef, avoid "_type" in type names and the suffix "_values" isn't much helpful.
```cpp
struct ring_signature
{
    ec_secret e;
    secret_list s;
};
```

The cryptic single letter variable names are acceptable within the internals of the implementation (only if they are helpful in correlating to some standard), but should be avoided in public interfaces. Assume people don't know what "e" or "s" mean and replace with informative names.

Reply:
I believe s is a list of lists, not the secret list.  it would be best to identify s_values_type as instead a secret_rings or equivalent to the key_rings.  Maybe public/private or other consistent naming thereof.

Reply:
Ideally I don't want people modifying the e and s values themselves. In that sense, they are internal values, however there is no defined serialization format. They are abstract variables with no given names in the paper.

Reply:
R_i_j, e_0 (or e), e_i_j, s_i_j, k_i <- these are the values used

Reply:
Okay, if the ring_sigature object has no public serialization it’s internals should remain hidden, and it should not be publicly constructible. To do this while maintaining essentially the same interface you can place the functions in a static class, define the ring_signature as a class in that context, give it a protected constructor only and expose its property methods and construction to only methods of that class.

Reply:
There's no defined serialization format, but still it needs to be serialized. So I'd need to access the internals in my code for saving the signature data.

Reply:
This means we would be defining a public serialization. In that case if the values s and e are not publicly meaningful they shouldn’t be exposed. Instead declare public construction from a byte array|chunk and provide from_data/to_data methods consistent with other serializable wallet/chain/config types and make e and s private internals. Basically if these properties aren’t publicly meaningful they shouldn’t be public. In fact it seems more reasonable that the signature just be a byte array or chunk, just like the existing ec signature, and perform the parsing internal to the one function that needs/understands it.

Reply:
I can invent a data serialization format in the format [ EC_SECRET ][ N:4 ][ M_0:4 ][ EC_SECRET ]...[ EC_SECRET ][ M_1:4 ]...[ M_(N - 1):4 ][ EC_SECRET ]...[ EC_SECRET ] which would be 32 + 4 + 4 * N + (M_0 * 32 + M_1 * 32 + ... + M_(N - 1) * 32) bytes long. How does that sound?

Reply:
This doesn’t look like a signature (ie contains secrets). It would seem all you need is [e][s].

Reply:
The serialization doesn’t have to expose publicly its encoding, it can be treated as opaque. Nobody using libbitcoin needs to be able to externally encode/decode the wire/store serialization of any object.